mmv1/third_party/terraform/acctest/resource_inventory_reader.go

```diff
@@ -0,0 +1,249 @@
+				servicePackageCache[metadata.Resource] = servicePackage
+				servicePkgCount++
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("error walking directory: %v", err)
+	}
+
+	// Mark cache as populated
+	cachePopulated = true
+
+	return nil
+}
+
+// GetCacheStats returns statistics about the metadata cache
+func GetCacheStats() (apiServiceNameCount int, servicePackageCount int) {
+	cacheMutex.RLock()
+	defer cacheMutex.RUnlock()
+
+	return len(apiServiceNameCache), len(servicePackageCache)
+}
+
+// GetAPIServiceNameForResource finds the api_service_name for a given resource name
+// If projectRoot is empty, it will attempt to find the project root automatically
+func GetAPIServiceNameForResource(resourceName string) string {
+	// Make sure cache is populated
+	if !cachePopulated {
+		if err := PopulateMetadataCache(); err != nil {
+			return "failed_to_populate_metadata_cache"
+		}
+	}
+
+	// Check cache
+	cacheMutex.RLock()
+	apiServiceName, found := apiServiceNameCache[resourceName]
+	cacheMutex.RUnlock()
+
+	if !found {
+		return "unknown"
+	}
+
+	return apiServiceName
+}
+
+// GetServicePackageForResourceType finds the service package for a given resource type
+// If projectRoot is empty, it will attempt to find the project root automatically
+func GetServicePackageForResourceType(resourceType string) string {
+	// Make sure cache is populated
+	if !cachePopulated {
+		if err := PopulateMetadataCache(); err != nil {
+			return "failed_to_populate_metadata_cache"
+		}
+	}
+
+	// Check cache
+	cacheMutex.RLock()
+	servicePackage, found := servicePackageCache[resourceType]
+	cacheMutex.RUnlock()
+
+	if !found {
+		return "unknown"
+	}
+
+	return servicePackage
+}
+
+// getServicesDir returns the path to the services directory
+// If projectRoot is provided, it will use that as the base
+// Otherwise, it will attempt to find the project root, and if that fails,
+// it will use environment variables or fall back to the current directory
+func getServicesDir(projectRoot string) (string, error) {
+	// If project root is provided, use it
+	if projectRoot != "" {
+		servicesDir := filepath.Join(projectRoot, "google-beta", "services")
+		if _, err := os.Stat(servicesDir); err == nil {
+			return servicesDir, nil
+		}
+		return "", fmt.Errorf("services directory not found at %s", servicesDir)
+	}
+
+	// Try to find project root
+	root, err := findProjectRoot()
+	if err == nil {
+		servicesDir := filepath.Join(root, "google-beta", "services")
+		if _, err := os.Stat(servicesDir); err == nil {
+			return servicesDir, nil
+		}
+	}
+
+	// Try environment variable
+	if envRoot := os.Getenv("TERRAFORM_PROVIDER_ROOT"); envRoot != "" {
+		servicesDir := filepath.Join(envRoot, "google-beta", "services")
+		if _, err := os.Stat(servicesDir); err == nil {
+			return servicesDir, nil
+		}
+	}
+
+	// Last resort: try relative to current directory
+	currentDir, err := os.Getwd()
```

Comment:
Is the last try for relative directory enough? Are the other options (finding project root, environment variable) necessary?

Reply:
I would like to keep find project root as i think this is the most adaptable? but I am not sure, yes the relative could be good enough.